File: services/audit-compliance-service/src/models/data_classification.rs
use std::str::FromStr;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ClassificationLevel {
    Public,
    Confidential,
    Pii,
}

impl ClassificationLevel {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Public => "public",
            Self::Confidential => "confidential",
            Self::Pii => "pii",
        }
    }

    pub fn requires_masking(self) -> bool {
        matches!(self, Self::Pii | Self::Confidential)
    }
}

impl FromStr for ClassificationLevel {
    type Err = String;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        match value {
            "public" => Ok(Self::Public),
            "confidential" => Ok(Self::Confidential),
            "pii" => Ok(Self::Pii),
            _ => Err(format!("unsupported classification level: {value}")),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClassificationCatalogEntry {
    pub classification: ClassificationLevel,
    pub description: String,
}

impl ClassificationCatalogEntry {
    pub fn new(classification: ClassificationLevel, description: &str) -> Self {
        Self {
            classification,
            description: description.to_string(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnomalyAlert {
    pub id: uuid::Uuid,
    pub title: String,
    pub description: String,
    pub severity: String,
    pub detected_at: DateTime<Utc>,
    pub correlation_key: String,
    pub linked_event_id: uuid::Uuid,
    pub recommended_action: String,
}

File: services/audit-compliance-service/internal/models/data_classification.go
// Data-classification + anomaly-alert types.
//
// Mirrors `services/audit-compliance-service/src/models/data_classification.rs`
// 1:1.
package models

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

// ClassificationLevel enumerates the three-tier data-classification
// vocabulary used everywhere in the audit ledger.
type ClassificationLevel string

const (
	ClassificationPublic       ClassificationLevel = "public"
	ClassificationConfidential ClassificationLevel = "confidential"
	ClassificationPii          ClassificationLevel = "pii"
)

// RequiresMasking returns true for confidential + pii (mirror of the
// Rust `requires_masking` helper).
func (c ClassificationLevel) RequiresMasking() bool {
	return c == ClassificationConfidential || c == ClassificationPii
}

// ParseClassificationLevel mirrors `ClassificationLevel::from_str`.
func ParseClassificationLevel(value string) (ClassificationLevel, error) {
	switch value {
	case "public":
		return ClassificationPublic, nil
	case "confidential":
		return ClassificationConfidential, nil
	case "pii":
		return ClassificationPii, nil
	default:
		return "", fmt.Errorf("unsupported classification level: %s", value)
	}
}

// AnomalyAlert mirrors the Rust struct.
type AnomalyAlert struct {
	ID                  uuid.UUID `json:"id"`
	Title               string    `json:"title"`
	Description         string    `json:"description"`
	Severity            string    `json:"severity"`
	DetectedAt          time.Time `json:"detected_at"`
	CorrelationKey      string    `json:"correlation_key"`
	LinkedEventID       uuid.UUID `json:"linked_event_id"`
	RecommendedAction   string    `json:"recommended_action"`
}

// ClassificationCatalogEntry mirrors the Rust struct of the same name.
type ClassificationCatalogEntry struct {
	Classification ClassificationLevel `json:"classification"`
	Description    string              `json:"description"`
}

// CollectorStatus is shared between the policy + events handlers; the
// Rust code keeps it under `models::policy::CollectorStatus`. We
// expose it here so both the read-model handler (`list_collectors`)
// and the overview handler can hold an instance without an extra
// alias.
type CollectorStatus struct {
	ServiceName  string     `json:"service_name"`
	Subject      string     `json:"subject"`
	Connected    bool       `json:"connected"`
	LastEventAt  *time.Time `json:"last_event_at"`
	BacklogDepth int32      `json:"backlog_depth"`
	Health       string     `json:"health"`
	NextPullAt   time.Time  `json:"next_pull_at"`
}
